### PR TITLE
Remove cluster from ApiClient

### PR DESCRIPTION
**Pull request by mellenburg**
-Circular dependencies are bad. This only worked because python doesnt evaluate until runtime
-Allows ApiClient to be generic
-Adds a convenience constructor to make the Api from a cluster object

### CODE REVIEW

**Comment by mellenburg** [comment]
@cmaloney This is ready for review

File: test_util/cluster_api.py [hunk 1]
```diff
@@ -244,9 +249,17 @@ def __init__(self, dcos_uri, masters, public_masters, slaves, public_slaves,
                 Note: user must be authenticated explicitly or call self.wait_for_dcos()
             ca_cert_path: (str) optional path point to the CA cert to make requests against
         """
+        # URI must include scheme
```
Comment (jgehrcke) [inline test_util/cluster_api.py]: In our case, that URI will _always_ be a URL. That's why I liked calling it a URL :-).

The fact that `URI must include scheme` actually is an artifact of this.

File: packages/dcos-integration-test/extra/test_applications.py [hunk 2]
```diff
@@ -142,13 +142,14 @@ def test_octarine_srv(cluster, timeout=30):
 
 def test_pkgpanda_api(cluster):
 
-    def get_and_validate_package_ids(node, uri):
-        r = cluster.get(node=node, path=uri)
+    def get_and_validate_package_ids(node, path):
```
Comment (cmaloney) [inline packages/dcos-integration-test/extra/test_applications.py]: :+1: to the function signature change

File: packages/dcos-integration-test/extra/test_applications.py [hunk 3]
```diff
@@ -142,13 +142,14 @@ def test_octarine_srv(cluster, timeout=30):
...
+        node_url = cluster.get_node_url(node)
```
Comment (cmaloney) [inline packages/dcos-integration-test/extra/test_applications.py]: the two calls here are a considerable API degredation from previously. Maybe cluster() needs to override / inject into the ClusterApi() a wrapper which makes it so that it will take either "host_url", _or_ if given the keyword parameter `node=` it will calculate host_url from that (both given or neither given being an error).

File: test_util/cluster_api.py [hunk 4]
```diff
@@ -263,39 +276,56 @@ def __init__(self, dcos_uri, masters, public_masters, slaves, public_slaves,
...
+
     def get_client(self, path, default_headers=None):
-        return test_util.helpers.ApiClient(
+        return test_util.helpers.ApiClient.from_cluster(
             cluster=self,
-            user=self.web_auth_default_user,
             api_base=path,
-            ca_cert_path=self.ca_cert_path,
-            default_headers=default_headers)
+            default_headers=None)
```
Comment (cmaloney) [inline test_util/cluster_api.py]: why is this not passing through `self.default_headers if default_headers is None else default_headers`?

Lets people set them to empty if they want / need, provides the same responses they would have gotten by hand-prefixing their requests.

File: test_util/helpers.py [hunk 5]
```diff
@@ -50,74 +49,50 @@ def authenticate(self, cluster):
...
         self.default_headers = default_headers
         self.ca_cert_path = ca_cert_path
 
-    def api_request(self, method, path, node=None, port=None, **kwargs):
+    @classmethod
+    def from_cluster(cls, cluster, api_base, default_headers=None):
+        if default_headers is None:
+            default_headers = {}
+        if cluster.auth_enabled and cluster.web_auth_default_user:
+            default_headers.update(cluster.web_auth_default_user.auth_header)
```
Comment (cmaloney) [inline test_util/helpers.py]: why not just copy over cluster.default_headers always?

File: test_util/helpers.py [hunk 6]
```diff
@@ -50,74 +49,50 @@ def authenticate(self, cluster):
...
-            port: see get_url()
+            host_url: override the client's default host URL
             **kwargs: any optional arguments to be passed to requests.request
         """
         headers = copy.copy(self.default_headers)
-        if self.cluster.auth_enabled and self.cluster.web_auth_default_user:
-            headers.update(self.cluster.web_auth_default_user.auth_header)
 
-        if self.ca_cert_path:
+        if self.ca_cert_path and 'verify' not in kwargs:
```
Comment (cmaloney) [inline test_util/helpers.py]: why does this get the verify part added now?

File: test_util/marathon.py [hunk 7]
```diff
@@ -9,7 +9,17 @@
 
 import test_util.helpers
 
+DEFAULT_API_BASE = '/marathon/v2'
```
Comment (cmaloney) [inline test_util/marathon.py]: a whole bunch of pods stuff is going to want slightly different stuff from /marathon/v2 I think (Although should confirm with that team).

File: test_util/marathon.py [hunk 8]
```diff
@@ -9,7 +9,17 @@
...
 TEST_APP_NAME_FMT = '/integration-test-{}'
+REQUIRED_HEADERS = {'Accept': 'application/json, text/plain, */*'}
+
+
+def update_headers(headers):
+    if headers:
```
Comment (cmaloney) [inline test_util/marathon.py]: `if headers is None`? This code seems complicated for what it does. Why isn't just a `headers.update(REQUIRED_HEADERS)` good enough?

File: packages/dcos-integration-test/extra/test_applications.py (hunk 3 above)
Comment (mellenburg) [inline packages/dcos-integration-test/extra/test_applications.py]: I thought wrapping a wrapper was bad but having to call cluster twice in one go is a little worse I now feel, so I will give it a shot

File: test_util/helpers.py (hunk 5 above)
Comment (mellenburg) [inline test_util/helpers.py]: Confused myself. Just copying cluster.default_headers is the right thing to do

File: test_util/helpers.py (hunk 6 above)
Comment (mellenburg) [inline test_util/helpers.py]: this allows people to make requests to ApiClient with verify=False or verify=my_other_custom_ca. Otherwise, the verify kwarg is always overridden, which makes for a bad requests wrapper

File: test_util/marathon.py (hunk 8 above)
Comment (mellenburg) [inline test_util/marathon.py]: the point of this function is just to handle the repeated case where default_headers=None so you don't have to rewrite:
```
if default_headers is None:
    default_headers = {}
```
but its not much of a savings

File: test_util/marathon.py (hunk 7 above)
Comment (mellenburg) [inline test_util/marathon.py]: Should there not be a default api version? I just figured the pods tests can then specifically ask for a marathon api with a different version

File: test_util/cluster_api.py (hunk 4 above)
Comment (mellenburg) [inline test_util/cluster_api.py]: Definitely a bug. Thanks for catching that.

File: test_util/helpers.py (hunk 6 above)
Comment (cmaloney) [inline test_util/helpers.py]: My point was it's an odd code change for _this_ PR, seems unrelated to the core goal of this PR and doesn't have a comment about why it changed in the code.

Comment (cmaloney) [inline test_util/helpers.py]: `# Set verify only if it isn't already set so callers can override the CA Cert`

File: test_util/marathon.py (hunk 7 above)
Comment (cmaloney) [inline test_util/marathon.py]: I'd rather have the api base be `/marathon`, people are familiar with `v2/apps` and friends as the API endpoints they call, rather than just "apps".

File: test_util/helpers.py [hunk 9]
```diff
@@ -48,85 +47,56 @@ def authenticate(self, cluster):
...
-        """
-        Args:
-            path: (str) URL path to request if self.api_base is set it will be prepended
-            node: (str) the hostname of the node to be requested from, if node=None,
-                then public cluster address (see environment DCOS_DNS_ADDRESS)
-            port: (int) port to be requested at. If port=None, the default port
-                for that given node type will be used
-        Returns:
-            fully-qualified URL string for this API
+    def wrapped_request(self, request_fn, *args, **kwargs):
```
Comment (mellenburg) [inline test_util/helpers.py]: @cmaloney What do you think about doing it this way?

File: test_util/cluster_api.py [hunk 10]
```diff
@@ -263,39 +276,76 @@ def __init__(self, dcos_uri, masters, public_masters, slaves, public_slaves,
...
-            cluster=self,
-            ca_cert_path=self.ca_cert_path,
-            user=self.web_auth_default_user)
+        marathon_client = test_util.marathon.Marathon(
+            default_host_url=self.dcos_url,
+            default_os_user=self.default_os_user,
+            default_headers=self.default_headers,
+            ca_cert_path=self.ca_cert_path)
+        # give spawned clients the node= request option
+        marathon_client.request_wrapper = self.request_wrapper
```
Comment (mellenburg) [inline test_util/cluster_api.py]: @cmaloney having to manually attach a new wrapper is a little clunky, but it still seems better than getting another class

**Comment by mellenburg** [comment]
@orsenthil PTAL

File: test_util/helpers.py [hunk 11]
```diff
@@ -48,85 +47,56 @@ def authenticate(self, cluster):
...
-
-    get = functools.partialmethod(api_request, 'get')
-    post = functools.partialmethod(api_request, 'post')
-    put = functools.partialmethod(api_request, 'put')
-    delete = functools.partialmethod(api_request, 'delete')
-    options = functools.partialmethod(api_request, 'options')
-    head = functools.partialmethod(api_request, 'head')
-    patch = functools.partialmethod(api_request, 'patch')
-    delete = functools.partialmethod(api_request, 'delete')
+        return self.request_wrapper(request_fn, *args, **kwargs)
```
Comment (orsenthil) [inline test_util/helpers.py]: I reviewed this piece too. The change is easy to understand. So, this is fine with me.

File: test_util/cluster_api.py [hunk 12]
```diff
@@ -263,39 +276,76 @@ def __init__(self, dcos_uri, masters, public_masters, slaves, public_slaves,
...
+                for that given node type will be used
+        Returns:
+            fully-qualified URL string for this API
+        """
+        if node in self.masters:
+            role = 'master'
+        elif node in self.all_slaves:
+            role = 'agent'
+        else:
+            raise Exception('Node {} is not recognized within the DC/OS cluster'.format(node))
```
Comment (orsenthil) [inline test_util/cluster_api.py]: Specific exception instead of a generic one?

File: test_util/cluster_api.py [hunk 13]
```diff
@@ -263,39 +276,76 @@ def __init__(self, dcos_uri, masters, public_masters, slaves, public_slaves,
...
+
+    def get_client(self, path):
+        new_client = test_util.helpers.ApiClient(
+            default_host_url=self.dcos_url,
             api_base=path,
             ca_cert_path=self.ca_cert_path,
-            default_headers=default_headers)
+            default_headers=self.default_headers)
+        # give spawned clients the node= reqest option
+        new_client.request_wrapper = self.request_wrapper
```
Comment (orsenthil) [inline test_util/cluster_api.py]: Is there better way for this? ( why we need to extend the client and a property attribute (line 341)with this additional method?

(this part is slightly confusing and bit difficult to determine the flow explicitly. The comments (which give details  about business logic ( something about node= options) is not indicative either).

File: test_util/cluster_api.py (hunk 12 above)
Comment (mellenburg) [inline test_util/cluster_api.py]: Which specific one would you prefer? I prefer generic Exceptions with specific messages until we actually are trying to catch the exception.

File: test_util/cluster_api.py (hunk 13 above)
Comment (mellenburg) [inline test_util/cluster_api.py]: The new_client has its own self.request wrapper from ApiClient, however, when we spawn this from a ClusterApi, we want the request signature to match ClusterApi. The difference in the two is that in ClusterApi, our request wrapper checks for the keyword 'node' and can translate that to a host_url before just calling out the ApiClient's request_wrapper. By setting the new_client request wrapper to be ClusterApi's request wrapper, they too can specify a node.

Does that explanation suffice in comment form or is there something you can suggest to restructure? I tried a few approaches and this is the cleanest yet
